hazelcast-sql/src/main/java/com/hazelcast/sql/impl/type/QueryDataType.java

```diff
@@ -336,18 +310,13 @@ public int hashCode() {
 
     @Override
     public boolean equals(Object o) {
-        // TODO: proper equals
```

Comment:
is this `equals` (and `hashCode`) reasonable for UDTs? They all have the same `converter` but different other attributes. If so, add a comment explaining that.

Reply:
`QueryDataType` is stored in `Mapping`, `TableField`, `MapIndexScanMetadata`, `ToConverters#CONVERTERS`, `AvroResolver#AVRO_TO_SQL`, `MetadataPortableResolver#PORTABLE_TO_SQL`, `HazelcastTypeUtils#CALCITE_TO_HZ`, `SqlPlanImpl.SelectPlan#partitionStrategyCandidates`, `CreateTopLevelDagVisitor#partitionStrategyCandidates` and various `Expression`s (including, but not limited to `(Uni/Bi/Vari)ExpressionWithType`s). I haven't detected any usage that doesn't tolerate changes in `hashCode()`/`equals()`, and I decided to implement proper `hashCode()`/`equals()` for custom types.

There is another problem, which was introduced after using null as the type metadata for some Avro types. Using empty string to indicate an unset variable is a bad idea in Java. I am also changing the default value of `objectTypeName` and `objectTypeMetadata` to null by making `QueryDataType` `Versioned`.